packages/dcos-integration-test/extra/test_metrics.py

```diff
@@ -405,3 +410,314 @@ def get_app_metrics(dcos_api_session, node: str, container_id: str):
+
+        wait_for_container_metrics_propogation()
+
+        get_containers = {
+            "type": "GET_CONTAINERS",
+            "get_containers": {
+                "show_nested": True,
+                "show_standalone": True
+            }
+        }
+
+        r = dcos_api_session.post('/agent/{}/api/v1'.format(agent_id), json=get_containers)
+        r.raise_for_status()
+        mesos_agent_containers = r.json()['get_containers']['containers']
+
+        assert len(mesos_agent_containers) == num_containers, 'Agent operator API should report '\
+            'exactly {} running containers'.format(num_containers)
+
+        def is_nested_container(container):
+            """Helper to check whether or not a container returned in the
+            GET_CONTAINERS response is a nested container.
+            """
+            return 'parent' in container['container_status']['container_id']
+
+        def check_tags(tags: dict, expected_tag_names: set):
+            """Assert that tags contain only expected keys with nonempty values."""
+            assert set(tags.keys()) == expected_tag_names
+            for tag_name, tag_val in tags.items():
+                assert tag_val != '', 'Value for tag "%s" must not be empty'.format(tag_name)
+
+        for container in mesos_agent_containers:
+            container_id = container['container_id']['value']
+
+            # Test that /containers/<id> responds with expected data.
+            container_id_path = '/containers/{}'.format(container_id)
+
+            if (is_nested_container(container)):
+                # Retry for 30 seconds for each nested container to appear.
+                # Since nested containers do not report resource statistics, we
+                # expect the response code to be 204.
+                @retrying.retry(stop_max_delay=30000)
+                def wait_for_container_response():
+                    response = dcos_api_session.metrics.get(container_id_path, node=agent_ip)
+                    assert response.status_code == 204
+                    return response
+
+                # For the nested container, we do not expect any container-level
+                # resource statistics, so this response should be empty.
+                assert not wait_for_container_response().json()
+
+                # Test that expected application metrics are present.
+                app_response = dcos_api_session.metrics.get('/containers/{}/app'.format(container_id), node=agent_ip)
+                assert app_response.status_code == 200, 'got {}'.format(app_response.status_code)
+
+                # Ensure all /container/<id>/app data is correct
+                assert 'datapoints' in app_response.json(), 'got {}'.format(app_response.json())
+
+                # We expect three datapoints, could be in any order
+                uptime_dp = None
+                for dp in app_response.json()['datapoints']:
```

Comment:
You mention 3 datapoints in the comment above, but there doesn't seem to be any logic or checking of this.  What's the importance of having 3?

Reply:
there will be two built-in app metrics (packets_received and packets_sent) and then the custom metric, in this case statsd_tester.time.uptime.

Reply:
This is an unfortunate artifact of a copy-paste job by yours truly. Fixed.